static/js/stream_popover.js

```diff
@@ -586,58 +625,25 @@ exports.register_topic_handlers = function () {
             return false;
         }
 
-        // The API endpoint for editing messages to change their
-        // content, topic, or stream requires a message ID.
-        //
-        // Because we don't have full data in the browser client, it's
-        // possible that we might display a topic in the left sidebar
-        // (and thus expose the UI for moving its topic to another
-        // stream) without having a message ID that is definitely
-        // within the topic.  (The comments in stream_topic_history.js
-        // discuss the tricky issues around message deletion that are
-        // involved here).
-        //
-        // To ensure this option works reliably at a small latency
-        // cost for a rare operation, we just ask the server for the
-        // latest message ID in the topic.
-        const data = {
-            anchor: "newest",
-            num_before: 1,
-            num_after: 0,
-            narrow: JSON.stringify([
-                {operator: "stream", operand: current_stream_id},
-                {operator: "topic", operand: old_topic_name},
-            ]),
-        };
+        const message_id = $("#move_topic_modal").data("message-id");
+
+        if (old_topic_name.trim() === new_topic_name.trim()) {
+            // We use `undefined` to tell the server that
+            // there has been no change in the topic name.
+            new_topic_name = undefined;
+        }
+
+        if (old_topic_name && select_stream_id) {
+            message_edit.move_topic_containing_message_to_stream(
+                message_id,
+                select_stream_id,
+                new_topic_name,
+                send_notification_to_new_thread,
+                send_notification_to_old_thread,
+            );
+            $("#move_topic_modal").modal("hide");
+        }
 
-        channel.get({
-            url: "/json/messages",
-            data,
-            idempotent: true,
-            success(data) {
-                const message_id = data.messages[0].id;
-
-                if (old_topic_name.trim() === new_topic_name.trim()) {
-                    // We use `undefined` to tell the server that
-                    // there has been no change in the topic name.
-                    new_topic_name = undefined;
-                }
-
-                if (old_topic_name && select_stream_id) {
-                    message_edit.move_topic_containing_message_to_stream(
-                        message_id,
-                        select_stream_id,
-                        new_topic_name,
-                        send_notification_to_new_thread,
-                        send_notification_to_old_thread,
-                    );
-                    $("#move_topic_modal").modal("hide");
-                }
-            },
-            error(xhr) {
-                show_error_msg(xhr.responseJSON.msg);
-            },
-        });
```

Comment:
I'm not sure this refactoring is actually advisable, in that now we have a user-facing network delay both when opening the modal and when submitting it.

I think it'd be not that hard to do a loading spinner with the chained 2-step network operation:
* Start the spinner at the beginning
* Stop the spinner in the error handler for the first request (keeping it going, by default, in the success handler)
* Stop the spinner in both success and error handlers for the second request.